Made `INVALID_DOC_ATTRIBUTES` lint deny by default

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/82730.

# Explanations

The `INVALID_DOC_ATTRIBUTES` lint was marked as "will become a hard error into the future" for quite some time so making it `deny` by default.

<del>Waiting on https://github.com/rust-lang/backtrace-rs/pull/524 for now.</del>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4959005673/jobs/8872565324) [(plain)](https://github.com/rust-lang/rust/commit/713b527ddddd8af3efbf7d286d88b9e56bad00a6/checks/13434715164/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling addr2line v0.19.0
error: this attribute can only be applied at the crate level
  --> library/std/src/../../backtrace/src/lib.rs:89:8
   |
89 | #![doc(html_root_url = "https://docs.rs/backtrace")]
   |
   |
   = note: read <https://doc.rust-lang.org/nightly/rustdoc/the-doc-attribute.html#at-the-crate-level> for more information
error: could not compile `std` (lib) due to previous error
Build completed unsuccessfully in 0:04:05
```

</details>

petrochenkov commented:
Blocked on https://github.com/rust-lang/backtrace-rs/pull/524.

GuillaumeGomez commented:
The backtrace PR got merged so isn't blocked anymore. :)

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5133118400/jobs/9238814409)
Build commit: 590af02d982816cfb2e4e43f8c54065e816581ce (`590af02d982816cfb2e4e43f8c54065e816581ce`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5133118400/jobs/9238814409"},"merge_sha":"590af02d982816cfb2e4e43f8c54065e816581ce"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5133118400/jobs/9238814409)
Build commit: 590af02d982816cfb2e4e43f8c54065e816581ce (`590af02d982816cfb2e4e43f8c54065e816581ce`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5133118400/jobs/9238814409"},"merge_sha":"590af02d982816cfb2e4e43f8c54065e816581ce"} -->

petrochenkov commented:
@craterbot check

Mark-Simulacrum commented:
Yielding crater to stable run: @craterbot cancel

Mark-Simulacrum commented:
@craterbot abort

petrochenkov commented:
@craterbot check

Enqueuing this PR for crater again, to run after [beta-170-4](https://crater.rust-lang.org/ex/beta-170-4).
@Mark-Simulacrum or do you mean other experiments cannot run together with beta at all? Due to improved crater parallelism or something.

GuillaumeGomez commented:
Interesting: all the regressions I saw are legit errors. Didn't expect so many people to ignore warnings though...

pheki commented:
For context, this warning first appeared in 1.52.0: https://github.com/rust-lang/rust/pull/82708

I looked into about 15 dependencies that are failing and about 10 "root result" failures, all dependencies I looked into were released before 1.52.0 and only one of the root result failures was for a crate released after 1.52.0. So I think most of the errors are for crates that never got that warning

petrochenkov commented:
`INVALID_DOC_ATTRIBUTES` seems to be reported for many different kinds of situations.

@GuillaumeGomez 
Could you extract all root errors from the logs and do `sort | uniq -c` on them so we can see statistics for specific errors?
@rustbot author

GuillaumeGomez commented:
Here are the results:

```
{'inline': 2180}
```

So the error is only triggered for the `doc(inline)` attribute.

<details><summary>Python code used to generate this result</summary>

```python
import os


def add_attr(errors, err, pos_start, len_elem):
    attr = err[pos_start:pos_start + len_elem]
    if attr not in errors:
        errors[attr] = 1
    else:
        errors[attr] += 1


def add_errors(errors, fpath):
    print("checking {}".format(fpath))
    with open(fpath, "r", encoding="utf8") as f:
        content = f.read()
    lines = content.split("\n")
    i = 0
    while i < len(lines):
        if lines[i] == "[INFO] [stdout] error: this attribute can only be applied to a `use` item":
	    # Now we extract the target attribute triggering this error.                                                                                                                                           
            err = lines[i + 3].split("|", 1)[1]
            sub_err = lines[i + 4].split("|", 1)[1]
            # We identify it from the "^" positions.                                                                                                                                                               
            pos_start = sub_err.index("^")
            len_elem = 0
            while len_elem + pos_start < len(sub_err):
                if sub_err[len_elem + pos_start] != "^":
                    # End of the "^", we can now add it to the errors.                                                                                                                                             
                    add_attr(errors, err, pos_start, len_elem)
                    pos_start = pos_start + len_elem
                    len_elem = 0
                    while pos_start < len(sub_err) and sub_err[pos_start] != "^":
                        pos_start += 1
                else:
                    len_elem += 1

            if len_elem != 0:
                add_attr(errors, err, pos_start, len_elem)
            i += 3
        i += 1


def main():
    errors = dict()
    for root, dirs, files in os.walk("."):
        for f in files:
            if f.startswith("try#"):
                add_errors(errors, os.path.join(root, f))
    print(errors)
```

</details>

petrochenkov commented:
>So the error is only triggered for the `doc(inline)` attribute.

So I guess we need to keep the `doc(inline)` case as a lint, and switch everything else to hard errors.
@rustbot author

GuillaumeGomez commented:
Or we can make it an error in edition 2024 for all attributes? No need for special case like that.

petrochenkov commented:
Maybe, I don't have much stake here, just want to avoid mass breakage.
(Edition specific behavior is also special casing though, and not a temporary one like deprecation lints.)

GuillaumeGomez commented:
Since we now have access to the 2024 edition internally, I was able to make it emit a warning for edition < 2024 and an error for edition >= 2024.

With this I think it's ready to go.

bors commented:
:umbrella: The latest upstream changes (presumably #116688) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Now that the 2024 edition was officialized, I think we can move forward with this PR.

r? @notriddle

GuillaumeGomez commented:
That's a very good point. I think adding a lint `#[doc(allow_unknown)]` would be a good way to circumvent this problem. Based on this (and if the team agrees), do you want it to be done in this PR or in another one?

Manishearth commented:
Let's let people discuss more, and probably make a separate PR linking to this one with a new FCP.

As a lint change it doesn't technically need an edition either but we could still do that to be nice, I believe the lint infra has machinery for "deny in next edition"

GuillaumeGomez commented:
Discussion about it happened [here](https://rust-lang.zulipchat.com/#narrow/stream/393423-t-rustdoc.2Fmeetings/topic/meeting.2006-11-2023/near/400607208) on zulip. The output of the conversation was that turning it into a hard error starting the 2024 edition is preferable as you can maintain MSRV with `cfg_attr(..., doc(...))`. However, as not everyone was present, I'll let others confirm whether or not they're okay with this approach before we continue.

Manishearth commented:
No, I disagree.  I already attempted to address this line of thinking in my comment:


> This is how most Rust features work; of course, however doc stuff tends to be secondary to everything else: I'd really hate to have to bump MSRV just to get prettier docs, especially when the error seems to affect the regular rustc build as well.

I do not think crates should have to:

 - Bump MSRV
 - Add a `nightly` feature (most crates don't have one! Also most people depending on the crate won't enable it)
 - Use `cfg(docsrs)` (makes regular docs not work)

to opt in to new _doc_ features; and we do want people to be able to play with these features as soon as they're available on stable. Each of these seems like a lot to give up for just better docs

GuillaumeGomez commented:
Well, putting it to the next meeting agenda then.

bors commented:
:umbrella: The latest upstream changes (presumably #119146) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
After discussion with the team, it was decided to keep it as a lint but to make it deny by default instead.

GuillaumeGomez committed:
Make `invalid_doc_attributes` deny by default

GuillaumeGomez commented:
Updated! It is now still a lint but deny by default.

Manishearth commented:
@rfcbot resolve multi-version

bors commented:
:umbrella: The latest upstream changes (presumably #121370) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #121790) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez committed:
Update ui tests

GuillaumeGomez commented:
@bors r=rustdoc rollup

bors commented:
:pushpin: Commit 8e817af3ae955aec532eba3d5eaaaf6621b700fa has been approved by `rustdoc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=rustdoc 8e817af3ae955aec532eba3d5eaaaf6621b700fa -->
<!-- homu: {"type":"Approved","sha":"8e817af3ae955aec532eba3d5eaaaf6621b700fa","approver":"rustdoc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
